src/python/pants/util/collections.py

```diff
@@ -63,3 +73,42 @@ def assert_single_element(iterable: Iterable[T]) -> T:
     return first_item
 
   raise ValueError(f"iterable {iterable!r} has more than one element.")
+
+
+E = TypeVar('E', bound='Enum')
+
+
+class EnumMatchError(ValueError):
+  """Issue when using match() on an enum."""
+
+
+class InexhaustiveMatchError(EnumMatchError):
+  """Not all values of the enum specified in the pattern match."""
+
+
+class UnrecognizedMatchError(EnumMatchError):
+  """A value is used that is not a part of the enum."""
+
+
+class Enum(StdLibEnum):
+
+  def all_values(self) -> ValuesView['Enum']:
+    return self.__class__.__members__.values()
```

Comment:
Does this `Enum` still work with option types like was implemented in #7304? Are there plans to do that?

Reply:
(This is also what the use of the `ChoicesMixin` in `objects.py` ensures right now -- automatically)

Reply:
This `Enum` works with our parser out of the box. Ran this little experiment:

```diff
diff --git a/src/python/pants/bin/pants_runner.py b/src/python/pants/bin/pants_runner.py
index 1a04ceb92..e8a005214 100644
--- a/src/python/pants/bin/pants_runner.py
+++ b/src/python/pants/bin/pants_runner.py
@@ -37,7 +37,7 @@ class PantsRunner(ExceptionSink.AccessGlobalExiterMixin):
     return not frozenset(self._args).isdisjoint(self._DAEMON_KILLING_GOALS)
 
   def _enable_rust_logging(self, global_bootstrap_options):
-    levelname = global_bootstrap_options.level.upper()
+    levelname = global_bootstrap_options.level.value.upper()
     init_rust_logger(levelname, global_bootstrap_options.log_show_rust_3rdparty)
     setup_logging_to_stderr(logging.getLogger(None), levelname)
 
diff --git a/src/python/pants/init/logging.py b/src/python/pants/init/logging.py
index 792a568ae..101d5fbb7 100644
--- a/src/python/pants/init/logging.py
+++ b/src/python/pants/init/logging.py
@@ -62,7 +62,7 @@ def setup_logging_to_stderr(python_logger, level):
 
 def setup_logging_from_options(bootstrap_options):
   # N.B. quiet help says 'Squelches all console output apart from errors'.
-  level = 'ERROR' if bootstrap_options.quiet else bootstrap_options.level.upper()
+  level = 'ERROR' if bootstrap_options.quiet else bootstrap_options.level.value.upper()
   native = Native()
   return setup_logging(level, console_stream=sys.stderr, log_dir=bootstrap_options.logdir, native=native)
 
diff --git a/src/python/pants/option/global_options.py b/src/python/pants/option/global_options.py
index 68415721d..f9c30df06 100644
--- a/src/python/pants/option/global_options.py
+++ b/src/python/pants/option/global_options.py
@@ -20,6 +20,7 @@ from pants.option.errors import OptionsError
 from pants.option.optionable import Optionable
 from pants.option.scope import ScopeInfo
 from pants.subsystem.subsystem_client_mixin import SubsystemClientMixin
+from pants.util.collections import Enum
 from pants.util.objects import enum
 
 
@@ -128,7 +129,13 @@ class GlobalOptionsRegistrar(SubsystemClientMixin, Optionable):
     default_distdir_name = 'dist'
     default_rel_distdir = f'/{default_distdir_name}/'
 
-    register('-l', '--level', choices=['trace', 'debug', 'info', 'warn'], default='info',
+    class Log(Enum):
+      trace = "trace"
+      debug = "debug"
+      info = "info"
+      warn = "warn"
+
+    register('-l', '--level', type=Log, default=Log.info,
              recursive=True, help='Set the logging level.')
     register('-q', '--quiet', type=bool, recursive=True, daemon=False,
              help='Squelches most console output. NOTE: Some tasks default to behaving quietly: '
```

--

What we still need to improve is the error message when using a value not in the Enum. Currently, it will print when running `./pants -ldebug2`: 

> Exception message: 'debug2' is not a valid Log`.

Instead, we want:

> pants.option.errors.ParseError: Error applying type 'Log2' to option value 'debug2', for option '--level' in global scope: type check error in class Log2: Value 'debug2' must be one of: ['trace', 'debug', 'info', 'warn'].

I was planning on addressing this in a dedicated followup PR that teaches `parser.py` to work with the stdlib `enum.Enum` (and thus `collections.Enum`).